components/webext-storage/src/migration.rs

```diff
@@ -0,0 +1,335 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+use crate::error::*;
+use rusqlite::{Connection, OpenFlags, Transaction, NO_PARAMS};
+use serde_json::{Map, Value};
+use sql_support::ConnExt;
+use std::path::PathBuf;
+
+// Simple migration from the "old" kinto-with-sqlite-backing implementation
+// to ours.
+// Could almost be trivially done in JS using the regular public API if not
+// for:
+// * We don't want to enforce the same quotas when migrating.
+// * We'd rather do the entire migration in a single transaction for perf
+//   reasons.
+
+// The sqlite database we migrate from has a very simple structure:
+// * table collection_data with columns collection_name, record_id and record
+// * `collection_name` is a string of form "default/{extension_id}"
+// * `record_id` is `key-{key}`
+// * `record` is a string with json, of form: {
+//     id: {the record id repeated},
+//     key: {the key},
+//     data: {the actual data},
+//     _status: {sync status},
+//     last_modified: {timestamp},
+//  }
+// So the info we need is stored somewhat redundantly.
+// Further:
+// * There's a special collection_name "default/storage-sync-crypto" that
+//   we don't want to migrate. Its record_id is 'keys' and its json has no
+//   `data`
+
+// We don't enforce a real quota, but don't want to exceed the
+// sync server's BSO size and also just be generally sane - so stop at 1.5MB
+// and also 2k of keys
+const MAX_LEN: usize = 1_512_000;
+const MAX_KEYS: usize = 2048;
+
+// The struct we read from the DB.
+struct LegacyRow {
+    col_name: String, // collection_name column
+    record: String,   // record column
+}
+
+impl LegacyRow {
+    // Parse the 2 columns from the DB into the data we need to insert into
+    // our target database.
+    fn parse(&self) -> Option<Parsed<'_>> {
+        if self.col_name.len() < 8 {
+            log::trace!("collection_name of '{}' is too short", self.col_name);
+            return None;
+        }
+        if &self.col_name[..8] != "default/" {
+            log::trace!("collection_name of '{}' isn't ours", self.col_name);
+            return None;
+        }
+        let ext_id = &self.col_name[8..];
+        let mut record_map = match serde_json::from_str(&self.record) {
+            Ok(Value::Object(m)) => m,
+            Ok(o) => {
+                log::trace!("record value is json, but not an object: {}", o);
```

Comment:
We'll want to record this information for telemetry... Unless that happens later.

It's probably worth logging what happened (but not the data, of course) at a higher level than trace as well.